### PR TITLE
Phi dominance check

### PR DESCRIPTION
**Pull request by dneto0**
This is a workaround to https://github.com/KhronosGroup/SPIRV-Tools/issues/286 
It disables def-use dominance checks for OpPhi (variable,predecessor) arguments.

### CODE REVIEW

**Comment by dneto0** [comment]
I filed isseu https://github.com/KhronosGroup/SPIRV-Tools/issues/288 to track the enhancement that would enable the (proposed) disabled test.

**Comment by umar456** [comment]
Ahh, sorry about that. I had made a special case for OpPhi in my previous implementation but I didn't move to the new commit. This seems like a reasonable workaround until i implement a proper fix.
LGTM +2

**Comment by dneto0** [comment]
Rebased

File: test/Validate.SSA.cpp [hunk 1]
```diff
@@ -1172,6 +1173,67 @@ TEST_F(ValidateSSA, PhiUseDoesntDominateUseOfPhiOperandUsedBeforeDefinitionBad)
...
+%value       = OpPhi %boolt %false %entry %copy %if_true
+               OpReturn
+               OpFunctionEnd
+)";
+
+  CompileSuccessfully(str);
+  ASSERT_EQ(SPV_SUCCESS, ValidateInstructions()) << getDiagnosticString();
+}
+
+TEST_F(ValidateSSA, DISABLED_PhiVariableDefMustComeFromBlockDominatingThePredecessorBad) {
```
Comment (umar456) [inline test/Validate.SSA.cpp]: Hmm. I am sure this was the intention but I don't see a reference to this rule anywhere in the spec. Here is what the instruction definition says about the `Parent` `Variable` pair:

```
Operands are a sequence of pairs: (/Variable 1/, /Parent 1/ block), (/Variable
2/, /Parent 2/ block),... Each /Parent i/ block is the label of an immediate
predecessor in the CFG of the current block. A /Parent i/ block must not appear
more than once in the operand sequence. All /Variables/ must have a type
matching /Result Type/.
```

Comment (dneto0) [inline test/Validate.SSA.cpp]: It's a requirement for SSA that it work that way. 
The SPIR-V spec says the module must be in SSA form but doesn't spell out all the rules for SSA.
The best I can point to is the section on SSA form: https://www.khronos.org/registry/spir-v/specs/1.1/SPIRV.html#_a_id_ssa_a_ssa_form  especially at the end where it mentions the use of a Phi.

Also, in the universal validation rules under the SSA bullet it says "Uses in a phi-function in a loop may consume definitions in the loop that don’t dominate the use."  Which is true (and vague) but also leaves out the forward-only but divergent then reconvergent control flow case like in this example (diamond structure).

**Pull request merged**